bigtable/internal/testproxy/proxy.go

```diff
@@ -202,30 +213,71 @@ func (s *goTestProxyServer) CreateClient(ctx context.Context, req *pb.CreateClie
 	}
 	opts = append(opts, credOpts...)
 
-	c, err := bigtable.NewClient(ctx, req.ProjectId, req.InstanceId, opts...)
+	localCtx, cancel := context.WithCancel(context.Background())
+
+	c, err := bigtable.NewClient(localCtx, req.ProjectId, req.InstanceId, opts...)
 	if err != nil {
+		cancel()
 		return nil, stat.Error(codes.Internal,
 			fmt.Sprintf("%s: failed to create client: %v", logLabel, err))
 	}
 
-	s.btClient = c
-	s.clientID = req.ClientId // Might need to be stored in a map
-	s.appProfileID = req.AppProfileId
-	s.perOperationTimeout = req.PerOperationTimeout
+	if _, exists := s.clientIDs[req.ClientId]; exists {
+		cancel()
+		return nil, stat.Error(codes.AlreadyExists,
+			fmt.Sprintf("%s: ClientID already exists %v", logLabel, err))
+	}
+
+	if s.clientIDs == nil {
+		s.clientIDs = make(map[string]testClient)
+	}
+
+	s.clientIDs[req.ClientId] = testClient{
```

Comment:
Should lock be used to have exclusive update of the map?

Reply:
+1 I'd guess that CreateClient should be thread safe and so a lock is required here.

Reply:
[facepalm] Yep, I completely forgot about making this application threadsafe. I'll update the code accordingly.